extra/migrate_settings.rb

```diff
@@ -1,108 +1,199 @@
+    ensure
+      migration_state.close unless migration_state.nil?
+    end
+  end
+
+  class Migration
+    attr_reader :working_dir_path
+
+    class << self
+      def inject_migrations_instance(migrations)
+        @@migrations = migrations
       end
+
+      def inherited(subclass)
+        @@migrations.register_migration(subclass)
+      end
+    end
+
+    def initialize(working_dir_path)
+      @working_dir_path = working_dir_path
+    end
+
+    def migration_name
+      underscore(self.class.name)
+    end
+
+    def migration_dir
+      File.join(working_dir_path, migration_name)
+    end
+
+    def src_dir
+      File.join(migration_dir, "src")
+    end
+
+    def dst_dir
+      File.join(migration_dir, "dst")
+    end
+
+    def underscore(src)
+      src = src.gsub(/::/, '/')
+      src = src.gsub(/([A-Z]+)([A-Z][a-z])/,'\1_\2')
+      src = src.gsub(/([a-z\d])([A-Z])/,'\1_\2')
+      src = src.tr("-", "_")
+      src.downcase
     end
   end
 
-  # Rename the port to whichever is correct
-  if output[:settings].keys.include?(:port)
-    if output[:settings].keys.include?(:ssl_certificate)
-      output[:settings][:https_port] = output[:settings].delete(:port)
-    else
-      output[:settings][:http_port] = output[:settings].delete(:port)
+  class Migrator
+    attr_reader :working_dir_path, :migrations_dir_path, :config_src_dir, :result_dir_path
+
+    def initialize(working_dir_path, migrations_dir_path, config_src_dir, migrations)
+      @working_dir_path = working_dir_path
+      @migrations_dir_path = migrations_dir_path
+      @config_src_dir = config_src_dir
+      @migrations = migrations
+      @result_dir_path = File.join(working_dir_path, 'result')
+    end
+
+    def load_migrations(migrations_dir_path)
+      migration_files = Dir.glob(File.join(migrations_dir_path, "*.rb")).sort
+      ::Proxy::Migration.inject_migrations_instance(@migrations)
+      migration_files.each { |file| require file }
+    end
+
+    def migrate
+      raise "Migrations dir '#{migrations_dir_path}' doesn't exist" unless Dir.exist?(migrations_dir_path)
+      raise "Settings dir '#{config_src_dir}' doesn't exist" unless Dir.exist?(config_src_dir)
+
+      puts "using: \n config files dir: #{config_src_dir},\n working dir: #{working_dir_path},\n migrations dir: #{migrations_dir_path},\n migrations state path: #{@migrations.migration_state_file_path}"
+
+      FileUtils.mkdir_p(result_dir_path)
+      load_migrations(migrations_dir_path)
+
+      executed_migrations = []
+      migrations = @migrations.new_migrations
+      migrations.each do |migration|
+        m = migration.new(working_dir_path)
+
+        create_migration_dirs(m.src_dir, m.dst_dir)
+        if migration == migrations.first
+          copy_original_configuration(m.src_dir)
+        else
+          copy_previous_migration_results(executed_migrations.last.dst_dir, m.src_dir)
+        end
+
+        m.migrate
```

Comment:
In that case you need to encapsulate this line bellow

    @migrations.persist_migrations_state(result_dir_path)

into `ensure` block. Because if an error occurs, all migrations that finished cleanly before the exception occured will never be persisted ergo executed multiple times once user resubmit the process.

I am still for catching/rethrowing the error and printing some *nice message* to the user how to proceed. Stacktrace is maybe nice for engineer, but not for end user.